### ISSUE TITLE
MacOS builds are broken due to missing Gstreamer plugins.

### ISSUE THREAD

**Issue by mukilan** [servo/servo#29653]
The [MacOS merge queue build](https://github.com/servo/servo/actions/runs/4761313063/jobs/8462444910) has failed with the following error:

```
[Errno 2] No such file or directory: '/usr/local/lib/libnice.dylib'
[Errno 2] No such file or directory: '/usr/local/lib/libnice.10.dylib'
[Errno 2] No such file or directory: '/usr/local/lib/gstreamer-1.0/libgstnice.dylib'
[Errno 2] No such file or directory: '/usr/local/lib/gstreamer-1.0/libgstwebrtc.dylib'
ERROR: could not find required GStreamer DLL: /usr/local/lib/libnice.dylib
ERROR: could not find required GStreamer DLL: /usr/local/lib/libnice.10.dylib
ERROR: could not find required GStreamer DLL: /usr/local/lib/gstreamer-1.0/libgstnice.dylib
ERROR: could not find required GStreamer DLL: /usr/local/lib/gstreamer-1.0/libgstwebrtc.dylib
```

Based on initial investigation, it looks like Homebrew's Gstreamer package was [recently updated](https://github.com/carlocab/homebrew-core/blob/5e81cb108b3ce0a8c1266fc95767c37e7b04e893/Formula/gstreamer.rb#L133) to disable building of libnice dependency which seems to be required for webrtc plugins. This might explain the above error.

Based on this [Github discussion in Hombrew](https://github.com/Homebrew/homebrew-core/pull/125996#discussion_r1166789425), it looks like this was deliberate and the recommendation is to use the officials binaries from GStreamer if webrtc is required.

**Comment by jdm** [comment]
In the short term, we can work around this by using the dummy media stack in CI.

**Comment by sagudev** [comment]
Can't we just use last homebrew version that still has libnice enabled? I believe it is 1.22 should be fine.

**Comment by mrobinson** [comment]
> Can't we just use last homebrew version that still has libnice enabled? I believe it is 1.22 should be fine.

Does homebrew have good support for installing old versions of most casks? Everything I've read seems to suggest some pretty intricate workarounds to install an old version, apart from some basic ones such as (python@3.9, python@3.10, etc).

**Comment by mukilan** [comment]
Right, I tried the versioning GStreamer last time the build broke due to the videoconvertscale plugin being split and my understanding is that Homebrew project prefers being on the latest version and the officially recommended strategy to create 'versioned' formulae is to maintain our own 'tap' (i.e repository of formulae definitions). I  did try creating my own 'tap' and it didn't work because of circular dependencies since at that point gstreamer was split into multiple formulae. Now that they've moved to a single 'monorepo'  Forumlua, things might be easier, but I guess it will be easier to move away from homebrew and provide prebuilt binaries for dependencies like we do on linux & windows.

Some relevant discussion can be found [here](https://github.com/Homebrew/brew/issues/8791#issuecomment-696643112)

**Comment by shirakaba** [comment]
I'm trying to build Servo on an M1 Mac. Happy to follow either the homebrew or Nix instructions (I happen to have nix-darwin set up), but unclear how to cut out gstreamer. I've tried each of the existing build instructions:

Build (via Homebrew):

```sh
# Via homebrew:
cd servo
brew bundle install --file=etc/homebrew/Brewfile
brew bundle install --file=etc/homebrew/Brewfile-build
pip install virtualenv
# Fails because the gst homebrew packages are missing

# Or, via Nix:
nix-shell etc/shell.nix
# Fails because it expects to find the gst packages installed
# under /opt/homebrew
```

If I comment-out the mentions of gst from `etc/homebrew/Brewfile` or `shell.nix`, it still fails to run:

 Running:
```sh
./mach build --dev && ./mach run tests/html/about-mozilla.html
# Fails with the following error:

[Errno 2] No such file or directory: '/opt/homebrew/lib/libnice.dylib'
[Errno 2] No such file or directory: '/opt/homebrew/lib/libnice.10.dylib'
[Errno 2] No such file or directory: '/opt/homebrew/lib/gstreamer-1.0/libgstnice.dylib'
[Errno 2] No such file or directory: '/opt/homebrew/lib/gstreamer-1.0/libgstwebrtc.dylib'
ERROR: could not find required GStreamer DLL: /opt/homebrew/lib/libnice.dylib
ERROR: could not find required GStreamer DLL: /opt/homebrew/lib/libnice.10.dylib
ERROR: could not find required GStreamer DLL: /opt/homebrew/lib/gstreamer-1.0/libgstnice.dylib
ERROR: could not find required GStreamer DLL: /opt/homebrew/lib/gstreamer-1.0/libgstwebrtc.dylib
```

**Comment by sagudev** [comment]
Well our mach script has some paths for gstreamer hard coded, which I tried to fix in my branch, but then I get different linking issues or missing libs errors and debbuging is hard for me as I do not have mac (only GH actions). You can take a look at https://github.com/sagudev/servo/tree/try-mac.

What I wanted to achieve is that you can provide GSTREAMER env variable that would contain gstreamer path, but the problem arose from mising dynlibs in runtime, which could be solved with otool (we alredy have same of otool logic in mach package command).

**Comment by mukilan** [comment]
@sagudev fyi, you might want to hold off on this. I've been looking into consuming official GStreamer package as well and at present I am able to build, run and package servo. The pending work is some refactoring across the different files and investigating an intermittent error message from dylib that I occasionally see when running servo about 2 different libraries for glib (from official framework and from homebrew paths) which I'm investigating.

**Comment by mukilan** [comment]
@shirakaba , to cut out gstreamer while building Servo, you can pass '--media-stack=dummy' to `./mach build` invocation.

**Comment by shirakaba** [comment]
@mukilan Thanks! That worked and I now have a handsome doge for my efforts!

<img width="512" alt="image" src="https://user-images.githubusercontent.com/14055146/235320104-ed078c04-1909-4287-a55a-e248ede1e936.png">

**Comment by mukilan** [comment]
@jdm , do you know what was the intention behind [copying ](https://github.com/servo/servo/blob/4339b3bab494e477b502f82137bd360a963ec2be/python/servo/build_commands.py#L794)both macos_dylibs and macos_plugins to `target/release` in the `./mach build` [step](https://github.com/servo/servo/blob/4339b3bab494e477b502f82137bd360a963ec2be/python/servo/build_commands.py#L633)?

The logic that we have today seems broken to me because 1) we don't copy the transitive dependencies of the gst dylibs and plugins and 2) only the plugins are actually [loaded ](https://github.com/servo/servo/blob/4339b3bab494e477b502f82137bd360a963ec2be/components/servo/lib.rs#L183) from servo binary relative path i.e   `target/release` . The other dependencies are still loaded from System GStreamer path (`/usr/local/Cellar/..`) because the `servo` binary and the copied dependencies still contain absolute install names, since we don't [patch the install names](https://github.com/servo/servo/blob/4339b3bab494e477b502f82137bd360a963ec2be/python/servo/package_commands.py#L140-L156) to be `@executable_path`-relative until the [packaging step](https://github.com/servo/servo/blob/4339b3bab494e477b502f82137bd360a963ec2be/python/servo/package_commands.py#L347).

Was the intention behind copying dependencies during build phase to ensure that ./mach run and other post build commands are run using a setup that is very similar to the final packaged dmg? Do you suggest we fix the issues with the build step above or just not copy dylibs during `./mach build` (and instead use system paths & standard GST environment variables where possible using the [set_run_env](https://github.com/servo/servo/blob/4339b3bab494e477b502f82137bd360a963ec2be/python/servo/command_base.py#L567) method)?

**Comment by jdm** [comment]
The intention was to create an environment that replicated the packaged builds, since we had recurring issues with writing media code that worked in local builds but not in nightly builds. For that reason, ny preference would be to fix the build step, rather than reverting to using the entire local gstreamer installation.

**Comment by mukilan** [comment]
Makes sense. Thanks for clarification, @jdm !
